### ISSUE TITLE
Warn when null-testing patterns are used on value types

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#54653]
**Version Used**: f32db1d1dc8bb41581582491e1f121c5ba2b8fca

**Steps to Reproduce**: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4GIYFcA2OLA4CmABETAcQLABQtAAgAwn0CMAdADKwCOA3LVoA3CACcSAdygIA5kQAuAZxIBeEjCISSAdWlz5AbQC6JAN4kERAGYRc8uOs0AKAJRmSABVEB7AA6qSACIIQJIAX3CBGigrEicpWQVFdgAxKFFFeQB5UQARa1sceVcSKGVzMLcAeiqyAA9fIgBjeSIEEghJMRhYGVpTWhIhyT0k1PTMnPybO1d2Lz92ABVvAGV5UV7XKLDBaNj40aVxjOy8gtm3MvcIgE9q2qIG5tb2zolu3v7B4dv5n18yzWGy2Lh2e3oAGYSJlRFgWjojt8aMMWNC2MwFv5zPo+DCFHjdjQwkA===)

```cs
#nullable enable

using System.Linq;

var widgets = new Widget[] { default, new() { Prop = "a" } };
if (widgets.FirstOrDefault() is { }) // expected a warning
{
    widgets.FirstOrDefault().Prop.ToString();
}

if (widgets.FirstOrDefault() is { } y) // expected a warning
{
    y.Prop.ToString();
}

public struct Widget
{
    public string Prop { get; set; }
}
```

Note that the codegen for the above patterns simply turns into `true`. I was actually really surprised that the `is { }` case didn't produce any warning, since the pattern is always `true`. I was less surprised about the absence of a warning on `is { } y`, but I'd like to argue in favor of producing a warning for it here.

When no warning is given in the above patterns, the reader may be mislead into thinking the pattern is performing some test on the input, when it's not doing any test at all. It seems like such a warning should be given whether or not a variable is also declared in the pattern. If the intent is to simply declare a variable without doing any tests, the user should adjust the code to use `is var y`, for example.

I think this would be a good warning wave candidate. It could potentially be expanded into warning on "vacuous" patterns in more places, although we should also account for use cases like for `a && b is var y`, where we want to avoid evaluating `b` unless `a` is true.

**Comment by jaredpar** [comment]
> I was actually really surprised that the is { } case didn't produce any warning, since the pattern is always true

I agree this is surprising but it is actually by explicit design. I made the exact same argument here but the LDM decision was that these should not produce any warnings. I can't remember the exact rationale but perhaps @agocke does. 

This is also why I prefer `is object` over `is {}`. The latter will warn when used on a type which cannot be `null` while the former will not. 

To take this change we'd need LDM to agree to change the spec here.
